src/number/streaming.rs

```diff
@@ -1578,6 +1663,79 @@ where
   Ok((i, float))
 }
 
+/// Recognizes floating point number in text format and returns a f64.
+/// This only handles non-finite (special) values.
+pub fn double_nonfinite<T, E: ParseError<T>>(input: T) -> IResult<T, f64, E>
+where
+  T: Slice<RangeFrom<usize>> + Slice<RangeTo<usize>>,
+  T: Clone + Offset,
+  T: InputIter + InputLength + InputTake + crate::traits::ParseTo<i32>,
+  <T as InputIter>::Item: AsChar,
+  <T as InputIter>::IterElem: Clone,
+  T: InputTakeAtPosition,
+  <T as InputTakeAtPosition>::Item: AsChar,
+  T: AsBytes,
+  T: for<'a> Compare<&'a [u8]>,
+{
+  let (i, sign) = sign(input.clone())?;
+  let (mut double, count) = if case_insensitive_cmp(i.as_bytes(), b"nan") {
+    (f64::NAN, 3)
+  } else if i.input_len() >= 8 && case_insensitive_cmp(i.as_bytes(), b"infinity") {
+    (f64::INFINITY, 8)
+  } else if case_insensitive_cmp(i.as_bytes(), b"inf") {
+    (f64::INFINITY, 3)
+  } else {
+    return Err(Err::Error(E::from_error_kind(input, ErrorKind::Float)));
+  };
+
+  if !sign {
+    double = -double;
+  }
+
+  if i.input_len() <= count {
```

Comment:
that not possible

Reply:
even worse I think this make it wrong cause "infinity"  IS complete no ? this mean that if would fail if input is just what we parse.

Reply:
> even worse I think this make it wrong cause "infinity" IS complete no ? this mean that if would fail if input is just what we parse. `if i.input_len() < count` is what should be write no ? but again, this check is not needed

I believe the check is needed, but I wasn't sure about `i.input_len() < count` (which is why I asked originally), so that sounds right.

Do we have to consider a case where we parse `b"inf"`, and in the streaming API, we don't know if this is complete or incomplete. Because if the following bytes are `b"inity"`, we've produced an incorrect result. The same is true for the float case: if our input is `b"1.2345"` and we've parsed our entire input, we might actually have `b"1.234567"`. I'm not especially well-versed in the internal implementations of nom, so any feedback or guidance is **highly** appreciated.

Reply:
Oh now I understand, that a good point, I **think** we should not if I look at doc of [`double()`](https://docs.rs/nom/7.0.0/nom/number/streaming/fn.double.html) for streaming the test show:

```rust
assert_eq!(parser("11e-1"), Ok(("", 1.1)));
assert_eq!(parser("123E-02"), Ok(("", 1.23)));
assert_eq!(parser("123K-01"), Ok(("K-01", 123.0)));
assert_eq!(parser("abc"), Err(Err::Error(("abc", ErrorKind::Float))));
```

That would say we should return Ok if we parsed something even if it "could" be wrong latter. I don't think there is a consensus here, the problem is that return incomplete forever is not good too cause the end of user output could be "inf". In practice there is very few streaming parser and that often "utils" parser and "real" parser often use complete variant. I don't think I see something truly use the streaming power of nom parser cause... that a very difficult like we see it just for something that simple.

So I think we would not return incomplete if we parse something.

Reply:
But I think we need @Geal input here

Reply:
I changed this to use the same logic as before, because I was having issues with the test suite and backwards compatibility should be favored. Previously, input like `"i"` would return an error, rather than `Incomplete`, and so I've made the special parser act like a complete parser:

https://github.com/Alexhuszagh/nom/blob/805e5be9d7e16a6fc0d86e3e6a16fd042d97fd3f/src/number/streaming.rs#L1507-L1526